hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java

```diff
@@ -6,139 +6,159 @@
  */
 package org.hibernate.internal.util.collections;
 
+import java.lang.reflect.Array;
 import java.util.ArrayDeque;
-import java.util.Deque;
-import java.util.Iterator;
+import java.util.Arrays;
+import java.util.NoSuchElementException;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
 
 /**
  * A general-purpose stack impl supporting null values.
+ * <p>
+ * Most of the implementation was inspired by {@link ArrayDeque} methods
+ * and follows the same logic.
  *
  * @param <T> The type of things stored in the stack
  *
  * @author Steve Ebersole
  * @author Sanne Grinovero
+ * @author Marco Belladelli
  */
 public final class StandardStack<T> implements Stack<T> {
+	private T[] elements;
+	private int head = 0;
+	private int tail = 0;
 
-	private ArrayDeque<T> internalStack;
-	private static final Object NULL_TOKEN = new Object();
-
-	public StandardStack() {
+	@SuppressWarnings("unchecked")
+	public StandardStack(Class<T> type) {
+		elements = (T[]) Array.newInstance( type, 8 );
```

Comment:
There's a non-negligible change that the `StandardStack` might be constructed, but then nothing gets added.

So I think it would be best to allocate the Array lazily?

Reply:
N.B. when I wrote this comment I hadn't realized the implementation was based on a circular array. This one might get clearer if you read the other comments first.